<?xml version="1.0"?>
<robot name="mallard" xmlns:xacro="http://www.ros.org/wiki/xacro">
  <xacro:include filename="ballast.xacro"/>
  <xacro:include filename="thruster.xacro"/>

  <!-- Thrusters components & constants -->
  <xacro:property name="PI" value="3.1415926535897931"/>
  <xacro:property name="tr1" value=".02" />
  <xacro:property name="tl1" value=".1" />
  <xacro:property name="tr2" value=".047" />
  <xacro:property name="tl2" value=".045" />

  <!-- <material name="blue">
    <color rgba="0.0 0.0 0.8 1.0"/>
  </material>
  <material name="black">
    <color rgba="0.0 0.0 0.0 1.0"/>
  </material>
  <material name="red">
    <color rgba="0.8 0.0 0.0 1.0"/>
  </material>
  <material name="gray">
    <color rgba="0.8 0.8 0.8 1.0"/>
  </material> -->

  <!--  *** MACROS *** -->
  <!-- Thruster link -->
  <!-- <xacro:macro name="thruster_link" params="name xyz rpy">
      
      <joint name="${name}" type="revolute">
          <parent link="base_link"/>
          <child link="${name}"/>
          <origin xyz="${xyz}" rpy="${rpy}"/>
          <dynamics damping="1.0"/>
          <axis xyz="0 1 0"/>
          <limit effort="30" velocity="5.0" lower="-${PI}" upper="${PI}" />
      </joint>
      
      <link name="${name}">      
          <visual name="${name}_1">
              <geometry>
                  <cylinder length="${tl1}" radius="${tr1}"/>
              </geometry>
              <material name="blue">
                  <color rgba="0 0 0.8 1"/>
              </material>
          </visual>  
          <visual name="${name}_2">
              <origin xyz="0 0 ${tl2/4}"/>
              <geometry>
                  <cylinder length="${tl2}" radius="${tr2}"/>
              </geometry>
              <material name="black">
                <color rgba="0 0 0 1"/>
              </material>
          </visual>  
          <inertial>
              <mass  value="0.3"/>
              <inertia ixx="0.001"
                       ixy="0.0"
                       ixz="0.0"
                       iyy="0.001"
                       iyz="0.0"
                       izz="0.001"/>
          </inertial>  
      </link> 

      <gazebo reference="${name}">
          <material>Gazebo/Black</material>
      </gazebo>

      <buoyancy>
          <compensation>0.6</compensation>
      </buoyancy>
  </xacro:macro>  -->

  <!-- <link name="odom">
  </link> -->

  <link name="base_link">
      <!-- Channels (aluminum connecting elements) -->
      <visual name="r_channel_visual">
        <geometry> 
          <box size="0.450 0.038 0.038"/>
        </geometry>
        <material name="gray">
          <color rgba="0.8 0.8 0.8 1.0"/>
        </material>
        <origin  xyz="0 -0.15 0" /> 
      </visual>
      <collision name="mallard_box">
        <origin xyz="0 0 0.057"/>
        <geometry> 
          <box size="0.564 0.5 0.152"/>
        </geometry>
        <!-- <origin rpy="0 0 0" xyz="0 0 0"/>  -->
      </collision>

      <visual name="l_channel_visual">
        <geometry> 
          <box size="0.450 0.038 0.038"/>
        </geometry>
        <origin xyz="0 0.15 0" /> 
      </visual>

      <!-- Cylinders (acrylic tubes)-->
      <visual name="front_tube_visual">
        <geometry>
          <cylinder radius="0.057" length="0.5"/>
        </geometry>
        <origin xyz="0.225 0 0.076" rpy="1.5707 0 0"/>
      </visual>

      <visual name="rear_tube_visual">
        <geometry>
          <cylinder radius="0.057" length="0.5"/>
        </geometry>
        <origin xyz="-0.225 0 0.076" rpy="1.5707 0 0"/>
      </visual>
     
      <!-- chassis physics -->
      <inertial>
        <origin xyz="0 0 -0.019" rpy="0 0 0"/>
        <mass value="10.5"/>
        <inertia ixx="0.17" ixy="0.0" ixz="0.0" iyy="0.25" iyz="0.0" izz="0.40"/>
      </inertial>

      <!-- The <limit> tag indicates where the buoyancy force 
      begins to decrease when approaching the water surface -->
      <buoyancy>
          <!-- Compensation coefficient is buoyancy force to weight ratio. Mallard
               buoyancy force is 11.52kg, weight is10.5kg.   -->
          <compensation>1.2</compensation>
          <origin xyz= "0 0 0"/>
          <limit radius="0.1"/>
          <damping xyz="100 100 100" rpy="2 2 2"/>
      </buoyancy>

  </link>

  <!-- <joint name="odom_to_base_link" type="fixed">
    <origin xyz="0.5 0.0 0.0" rpy="0.0 0.0 0.0"/>
    <parent link="odom"/>
    <child link="base_link"/>
  </joint> -->


  <link name="lidar">
    <visual name="lidar_base">
      <geometry>
        <box size="0.06 0.061 0.048"/>
      </geometry>
      <material name="blue">
        <color rgba="0.0 0.0 0.8 1.0"/>
      </material>
      <origin xyz="0.0 0.0 0.0"/>
    </visual>
    <collision>
      <geometry>
        <box size="0.06 0.061 0.048"/>
      </geometry>
      <origin xyz="0.0 0.0 0.0"/>
    </collision>

    <visual name="optical_hood">
      <origin xyz="0 0 0.0425"/>
      <geometry>
        <cylinder radius="0.025" length="0.037"/>
      </geometry>
      <material name="blue">
        <color rgba="0.0 0.0 0.8 1.0"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0.0 0.0 0.048"/>
      <geometry>
        <cylinder radius="0.025" length="0.037"/>
      </geometry>
    </collision>

    <inertial>
      <mass value="0.3"/>
      <inertia ixx="0.00020" ixy="0.0" ixz="0.0" iyy="0.00020" iyz="0.0" izz="0.00015"/>
    </inertial>
  </link>

  <joint name="lidar_to_base_link" type="fixed">
    <parent link="base_link"/>
    <child link="lidar"/>
    <origin xyz="0.225 0 0.152"/>
  </joint>

  <!-- Ballast -->
  <xacro:ballast name="ballast_1" mass="0.14" xyz=" 0.225 -0.25 0.076" rpy="${PI/2} 0 0"/>
  <xacro:ballast name="ballast_2" mass="0.02" xyz="-0.225 -0.25 0.076" rpy="${PI/2} 0 0"/>

  <!-- Forward thrusters -->
  <xacro:thruster_link name="x_thr_left"   xyz="${(tl1/2)-0.188}  0.207 -0.041" rpy="0 ${-PI/2} 0"/>
  <xacro:thruster_link name="x_thr_right"  xyz="${(tl1/2)-0.188} -0.207 -0.041" rpy="0 ${-PI/2} 0"/>
  <!-- Side thrusters -->
  <xacro:thruster_link name="y_thr_left"   xyz=" 0.233 ${0.114-(tl1/2)} -0.041" rpy="0 ${PI/2} ${-PI/2}"/>
  <xacro:thruster_link name="y_thr_right"  xyz="-0.233 ${0.114-(tl1/2)} -0.041" rpy="0 ${PI/2} ${-PI/2}"/>

  <!-- FreeFloating plugin -->
  <gazebo>
    <plugin name="freefloating_gazebo_control" filename="libfreefloating_gazebo_control.so">
          <switchService>switch</switchService>
          <updateRate>100</updateRate>
          <link>base_link</link>

                <!-- THRUST CONTROL -->
                <!--    these two thrusters are actually orientable, so the map should not be hard-coded 
                        the thruster name is the name of the corresponding link --> 
                <thruster>
                    <effort>10</effort>
                    <name>x_thr_left</name>
                </thruster>

                <thruster>
                    <effort>10</effort>
                    <name>x_thr_right</name>
                </thruster>

                <thruster>
                    <effort>10</effort>
                    <name>y_thr_left</name>
                </thruster>

                <thruster>
                    <effort>10</effort>
                    <name>y_thr_right</name>
                </thruster>
                
                <!-- <map> describes where the force or torque is applied
                with respect to body_link frame. Needs at leats one of them otherwise plugin won't work  --> 
                <thruster>
                    <map>0 1 0 0 0 0</map>
                    <effort>10</effort>
                </thruster>

      </plugin>
  </gazebo>


</robot>
